From fe671a01877129270421a7942259ddab85788061 Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Mon, 17 Aug 2026 22:42:44 -0400
Subject: [PATCH 279/285] winewayland: Use offscreen rendering for layered
 client surfaces.

Alpha modifier protocol doesn't specify how blending works, so we are forced to do this :(
---
 dlls/winewayland.drv/wayland_surface.c | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/dlls/winewayland.drv/wayland_surface.c b/dlls/winewayland.drv/wayland_surface.c
index d675d34f4ad..1c2eaa05c32 100644
--- a/dlls/winewayland.drv/wayland_surface.c
+++ b/dlls/winewayland.drv/wayland_surface.c
@@ -1450,11 +1450,16 @@ static void wayland_client_surface_update_offscreen(struct wayland_client_surfac
 {
     HWND hwnd = surface->client.hwnd, toplevel = NtUserGetAncestor(hwnd, GA_ROOT);
     LONG offscreen = force_use_offscreen();
-    DWORD pid;
+    DWORD pid, flags;
+    BYTE alpha;
 
     if (NtUserGetWindowThread(toplevel, &pid) && pid != GetCurrentProcessId())
         offscreen = TRUE;
 
+    if (!offscreen && NtUserGetLayeredWindowAttributes(toplevel, NULL, &alpha, &flags) &&
+        (flags & LWA_ALPHA) && alpha != 0xff)
+        offscreen = TRUE;
+
     if (!offscreen && NtUserGetWindowRelative(hwnd, GW_CHILD))
         offscreen = needs_client_window_clipping(hwnd);
 
-- 
2.55.0
